### DISCUSSION TITLE
Projects where "async void" is an anti-pattern

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6172] [Language Ideas] [answered]
While `async void` makes sense for event handlers in WinForms applications (or more generally, in rich client applications), it doesn't make sense in other scenarios, particularly web/API projects. In fact, it's most likely a nasty bug in those projects. This becomes a problem when doing heavy refactoring; it's easy to miss a `void` - not change it to `Task` - with undesired behavior at runtime.

Is it possible to add an "info"-level message for `async void`, so that it can be hidden in rich client projects, and configured to warning/error severity in others, by users as they see fit? Or would this be an VS feature? Perhaps there is an analyzer for this? Any info would be appreciated.

**Comment by CyrusNajmabadi** [comment, marked as answer]
I believe there are analyzers that already do this.  i think it may be called vs-threading?  @AArnott ?

**Reply by TahirAhmadov** [reply]
Thanks, yes it's here, works perfectly:
https://www.nuget.org/packages/Microsoft.VisualStudio.Threading.Analyzers/
